Option: output sparse binary dose file instead of 3ddose in DOSXYZnrc

marenaud opened the pull request:
# Sparse binary 3ddose output
Just throwing this pull request up here to see if there is any interest. It was a fairly simple change to make.

## Use case
As part of my research, I use DOSXYZnrc to generate electron beamlets for mixed electron-photon optimisation. A full set of electron beamlets for 5 gantry angles for a single chest wall patient (15,000 beamlets total) with 3mm^3 voxels takes up 2.1 TB of disk space when they're stored as 3ddose files. Just the process of converting beamlets to a nicer format takes a long time. When stored as sparse binary files, the same set of beamlets takes ~14 GB. Being able to output directly to binary dose saves tons of space and time.

## The format
The format is similar the 3ddose format, except that voxels with 0 dose are omitted. The binary file is constructed as:

- x_num_voxels, y_num_voxels, z_num_voxels (3 * sizeof(i32) bytes)
- Voxel coordinates in x direction ((x_num_voxels + 1) * sizeof(f32) bytes)
- Voxel coordinates in y direction ((y_num_voxels + 1) * sizeof(f32) bytes)
- Voxel coordinates in z direction ((z_num_voxels + 1) * sizeof(f32) bytes)
- Number of nonzero voxels (1 * sizeof(i32) bytes)
- 0-indexed linearized voxel indices (num_nonzero * sizeof(i32) bytes)
- Non-zero dose values (num_nonzero * sizeof(f32) bytes)
- Uncert values for non-zero voxels (num_nonzero * sizeof(f32) bytes)

The format was chosen to allow codes reading in the binary dose to avoid having to read in uncertainties if they're not needed. Dose/uncert values are stored as floats to save space since there's no accumulation of floating point errors to worry about as this is a storage format.

I added a flag to toggle between 3ddose/binary dose output in the input file. Default value is 3ddose and the input file is kept backwards compatible for people who don't specify the flag.

## Downsides
- The binary files aren't as portable as text files, though this is becoming less and less true as little-endian architecture seem to have won the endianness war?
- To output the binary dose file efficiently, I added an array to hold indices of nonzero voxels in the write_dose routine, adds $MAXDOSE * sizeof(i32) bytes to RAM usage of DOSXYZnrc. This can be avoided but the routine to output the binary dose becomes less efficient (need to do more passes through the dose arrays).
- Uses ```access='stream'``` in fortran to write C-style binary files. Apparently this was added with fortran2003 so there could be compatibility issues with older compilers.
- May not really be needed by anyone else?

Let me know if there is interest to include this in the official distribution. I can fill in the gaps in the documentation and add a checkbox to the GUI. Otherwise I'll just keep this on my branch.

Cheers,
Marc

randlet commented:
Looks pretty cool Marc!   Did you investigate how well your 3ddose files compress? In egs_brachy we have an option to output gzipped 3ddose (and egsdat) files (via [gzstream](https://www.cs.unc.edu/Research/compgeom/gzstream/)) instead of text.  IIRC the gzipped files were roughly an order of magnitude smaller so not as impressive as your gains here.  If you have lots of zero dose voxels though you may get better compression than our typical case.

marenaud commented:
Hey Randy!
Quick test with a 200x60x200 phantom (dimensions arbitrary, just took the first input file I could find).
```
-rw------- 1 marc users 120M Jan 15 11:45 minidos_test.3ddose
-rw------- 1 marc users 343K Jan 15 11:47 minidos_test.3ddose.tar.gz
-rw------- 1 marc users  25K Jan 15 11:46 minidos_test.bindos
-rw------- 1 marc users  19K Jan 15 11:50 minidos_test.bindos.tar.gz
```

Not too bad! For my specific use case, I'd still want to use a binary file because everytime I do an optimisation run, I need to read in those 15000 dose files. As you can imagine, 3ddose files are extremely inefficient to parse compared to binary files.

randlet commented:
Thanks for posting your numbers!  Your technique is still a pretty nice gain over just gzipping.  

I wonder if you could make it more generally applicable by storing start and end indices (or block sizes) for blocks of non zero doses instead of storing the individual voxel indices directly  Something like:

1) x_num_voxels, y_num_voxels, z_num_voxels 
2) Voxel coordinates in x direction
3) Voxel coordinates in y direction 
4) Voxel coordinates in z direction 
~~5) 0-indexed linearized voxel indices~~
5) List of pairs of start and end indices (or size of block)  for contiguous non zero block of doses
6) Non-zero dose values
7) Uncert values for non-zero voxels

Worst case for both methods (all voxels non zero for yours, every other voxel non zero for mine) stores Nvox indices so I guess it would be fairly dependent on how the zero doses are distributed throughout your dose array.

rtownson commented:
I think this option for sparse binary output is long overdue! I've coded a similar solution for my own purposes before, due to the recombination of 3ddose files to be the slowest part of certain highly parallel simulations.

I think Randle's suggestion of storing pairs makes sense, since dose usually occurs in contiguous regions.

marenaud commented:
It would be interesting to profile the different methods vs their savings in space.

Doing blocks of indices would save space on the indices, so you'd be optimising the 1/3 of the file size that can be optimised (ignoring the header). The cost would be an increase in complexity for writing/reading those files. Having 3 neat blocks means you can read the entire dose content of the file with 3 freads (or a single fread since they're all 4 bytes but then there's more work to do in the background).

marenaud commented:
I'm sitting down and implementing the "RandyDos" file format now, will have some benchmarks soon!

marenaud commented:
OK so I implemented the [randydos file format](https://github.com/marenaud/EGSnrc/blob/feature-bin-sparse-dose-index-pairs/HEN_HOUSE/user_codes/dosxyznrc/dosxyznrc.mortran#L4431).

Here's some numbers for a monoenergetic 16 MeV electron beam incident on a the xy plane of a 128x128x128 phantom
```
105M Jan 15 21:12 dummy.3ddose
2.9M Jan 15 21:17 dummy.bindos
2.2M Jan 15 22:27 dummy.randydos

5.5M Jan 15 22:41 dummy.3ddose.tar.gz
2.0M Jan 15 22:41 dummy.bindos.tar.gz
1.8M Jan 15 22:41 dummy.randydos.tar.gz
```

I then made a python class which deserializes each of these files into the same format. I print some statistics for (basic) validation
```
Loading dummy.3ddose
Number of nonzero dose voxels: 249672
Average dose of nonzero dose voxels: 1.20855807419e-10
Average uncertainty of nonzero dose voxels: 0.20378229854

Loading dummy.bindos
Number of nonzero dose voxels: 249672
Average dose of nonzero dose voxels: 1.20855807421e-10
Average uncertainty of nonzero dose voxels: 0.203782298532

Loading dummy.randydos
Number of nonzero dose voxels: 249672
Average dose of nonzero dose voxels: 1.20855807421e-10
Average uncertainty of nonzero dose voxels: 0.203782298532
```

The number of index pairs was 33862, so 67724 ints to store instead of 249672, a nice reduction.

and then I went into ipython to timed their load times
```
In [5]: %timeit Dose.from_3ddose("dummy.3ddose")
1 loop, best of 3: 914 ms per loop

In [6]: %timeit Dose.from_bindos("dummy.bindos")
10 loops, best of 3: 25.1 ms per loop

In [7]: %timeit Dose.from_randydos("dummy.randydos")
10 loops, best of 3: 41 ms per loop
```

This is far from scientific though since I can't guarantee that my from_randydos method is the most efficient way to deserialize a randydos. See below for the code I wrote for each file type.

```python
"""Dose module to load and represent voxelized Monte Carlo dose data"""
import struct
import numpy as np

class Dose(object):
    """Voxelized Monte Carlo dose data"""
    def __init__(self, num_voxels, x_voxels, y_voxels, z_voxels, doses, uncerts):
        self.num_voxels = num_voxels
        self.x_voxels = x_voxels
        self.y_voxels = y_voxels
        self.z_voxels = z_voxels
        self.doses = doses
        self.uncerts = uncerts

    def print_dose_statistics(self):
        print("Number of nonzero dose voxels: {}".format(np.count_nonzero(self.doses)))

        avg_dose = np.average(self.doses[np.nonzero(self.doses)])
        print("Average dose of nonzero dose voxels: {}".format(avg_dose))

        avg_uncert = np.average(self.uncerts[np.nonzero(self.doses)])
        print("Average uncertainty of nonzero dose voxels: {}".format(avg_uncert))


    @classmethod
    def from_3ddose(cls, path):
        "Create Dose instance from 3ddose file"
        print("Loading {}".format(path))

        with open(path, "r") as dose_file:
            num_voxels = [int(i) for i in dose_file.readline().split()]
            x_pos = np.array(dose_file.readline().split(), dtype=np.float)
            y_pos = np.array(dose_file.readline().split(), dtype=np.float)
            z_pos = np.array(dose_file.readline().split(), dtype=np.float)

            dose_array = np.array(dose_file.readline().strip().split(), dtype=np.float)
            doses = np.reshape(dose_array, (num_voxels[2], num_voxels[1], num_voxels[0]))

            uncert_array = np.array(dose_file.readline().strip().split(), dtype=np.float)
            uncerts = np.reshape(uncert_array, (num_voxels[2], num_voxels[1], num_voxels[0]))

        return cls(num_voxels, x_pos, y_pos, z_pos, doses, uncerts)

    @classmethod
    def from_bindos(cls, path):
        "Create Dose instance from bindos file"
        print("Loading {}".format(path))

        with open(path, "rb") as binfile:
            # num_voxels (3 ints)
            vox_fmt = "=3i"
            data = binfile.read(struct.calcsize(vox_fmt))
            num_voxels = struct.unpack(vox_fmt, data)

            # x_voxels (num_voxels[0] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[0]+1)
            data = binfile.read(struct.calcsize(voxels_fmt))
            x_pos = struct.unpack(voxels_fmt, data)

            # y_voxels (num_voxels[1] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[1]+1)
            data = binfile.read(struct.calcsize(voxels_fmt))
            y_pos = struct.unpack(voxels_fmt, data)

            # z_voxels (num_voxels[2] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[2]+1)
            data = binfile.read(struct.calcsize(voxels_fmt))
            z_pos = struct.unpack(voxels_fmt, data)

            # number of nonzero dose voxels (1 int)
            nonzero_fmt = "=i"
            data = binfile.read(struct.calcsize(nonzero_fmt))
            num_nonzero = struct.unpack(nonzero_fmt, data)[0]

            # voxel indices of nonzero dose voxels (num_nonzero ints)
            voxel_indices_fmt = "={}i".format(num_nonzero)
            data = binfile.read(struct.calcsize(voxel_indices_fmt))
            voxel_indices = struct.unpack(voxel_indices_fmt, data)
            voxel_indices = np.array(voxel_indices, dtype=np.int)

            # nonzero dose voxels (num_nonzero floats)
            nonzero_doses_fmt = "={}f".format(num_nonzero)
            data = binfile.read(struct.calcsize(nonzero_doses_fmt))
            nonzero_doses = struct.unpack(nonzero_doses_fmt, data)
            nonzero_doses = np.array(nonzero_doses)

            # nonzero uncert voxels (num_nonzero floats)
            nonzero_uncerts_fmt = "={}f".format(num_nonzero)
            data = binfile.read(struct.calcsize(nonzero_uncerts_fmt))
            nonzero_uncerts = struct.unpack(nonzero_uncerts_fmt, data)
            nonzero_uncerts = np.array(nonzero_uncerts)

            doses = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])
            uncerts = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])

            doses[voxel_indices] = nonzero_doses
            uncerts[voxel_indices] = nonzero_uncerts

            doses = np.reshape(doses, (num_voxels[2], num_voxels[1], num_voxels[0]))
            uncerts = np.reshape(uncerts, (num_voxels[2], num_voxels[1], num_voxels[0]))

        return cls(num_voxels, x_pos, y_pos, z_pos, doses, uncerts)

    @classmethod
    def from_randydos(cls, path):
        "Create Dose instance from randydos file"
        print("Loading {}".format(path))

        with open(path, "rb") as randyfile:
            # num_voxels (3 ints)
            vox_fmt = "=3i"
            data = randyfile.read(struct.calcsize(vox_fmt))
            num_voxels = struct.unpack(vox_fmt, data)

            # x_voxels (num_voxels[0] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[0]+1)
            data = randyfile.read(struct.calcsize(voxels_fmt))
            x_pos = struct.unpack(voxels_fmt, data)

            # y_voxels (num_voxels[1] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[1]+1)
            data = randyfile.read(struct.calcsize(voxels_fmt))
            y_pos = struct.unpack(voxels_fmt, data)

            # z_voxels (num_voxels[2] + 1 floats)
            voxels_fmt = "={}f".format(num_voxels[2]+1)
            data = randyfile.read(struct.calcsize(voxels_fmt))
            z_pos = struct.unpack(voxels_fmt, data)

            # number of nonzero dose voxels (1 int)
            nonzero_fmt = "=i"
            data = randyfile.read(struct.calcsize(nonzero_fmt))
            num_nonzero = struct.unpack(nonzero_fmt, data)[0]

            # number of voxel blocks (1 int)
            num_block_fmt = "=i"
            data = randyfile.read(struct.calcsize(num_block_fmt))
            num_blocks = struct.unpack(num_block_fmt, data)[0]

            # Nonzero voxel blocks (2 * num_blocks ints)
            block_fmt = "={}i".format(2 * num_blocks)
            data = randyfile.read(struct.calcsize(block_fmt))
            blocks = struct.unpack(block_fmt, data)

            # nonzero dose voxels (num_nonzero floats)
            nonzero_doses_fmt = "={}f".format(num_nonzero)
            data = randyfile.read(struct.calcsize(nonzero_doses_fmt))
            nonzero_doses = struct.unpack(nonzero_doses_fmt, data)
            nonzero_doses = np.array(nonzero_doses)

            # nonzero uncert voxels (num_nonzero floats)
            nonzero_uncerts_fmt = "={}f".format(num_nonzero)
            data = randyfile.read(struct.calcsize(nonzero_uncerts_fmt))
            nonzero_uncerts = struct.unpack(nonzero_uncerts_fmt, data)
            nonzero_uncerts = np.array(nonzero_uncerts)

            doses = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])
            uncerts = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])

            # No idea if this is the fastest way to go from voxel blocks to voxel indices
            voxel_indices = np.zeros(num_nonzero, dtype=np.int)
            processed_voxels = 0
            for block_i in range(num_blocks):
                block_length = blocks[2*block_i+1] - blocks[2*block_i]
                voxel_indices[processed_voxels: processed_voxels + block_length] = np.arange(blocks[2*block_i], blocks[2*block_i+1])
                processed_voxels += block_length

            doses[voxel_indices] = nonzero_doses
            uncerts[voxel_indices] = nonzero_uncerts

            doses = np.reshape(doses, (num_voxels[2], num_voxels[1], num_voxels[0]))
            uncerts = np.reshape(uncerts, (num_voxels[2], num_voxels[1], num_voxels[0]))

            return cls(num_voxels, x_pos, y_pos, z_pos, doses, uncerts)

```

randlet commented:
Nice work!

```python
voxel_indices2 = []
for vstart, vend in zip(blocks[::2], blocks[1::2]):
    voxel_indices2.extend(range(vstart, vend))

voxel_indices3 = [vi for vstart, vend in zip(blocks[::2], blocks[1::2]) for vi in range(vstart, vend)]
```

Either of those solutions for getting your voxel indices would be a bit more idiomatic in Python (may or may not be quicker, vi3 is probably quickest) but you could also just fill your doses array rather than creating the intermediate voxel indices array:

```python
# not tested :p

doses = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])
uncerts = np.zeros(num_voxels[0] * num_voxels[1] * num_voxels[2])

processed = 0
for block_start, block_end in zip(blocks[::2], blocks[1::2]):
    block_size = block_end - block_start
    doses[block_start:block_end] = nonzero_doses[processed:processed + block_size]
    uncerts[block_start:block_end] = nonzero_doses[processed:processed + block_size]
    processed += block_size

```

That said, you need to be cautious extrapolating the performance of Python to (Fortran|C|C++|etc)!

marenaud commented:
I'll test the performance, my logic was to use as much of numpy as possible over python loops because of the speed difference, but that could be premature optimisation!

marenaud commented:
[C++11 version](https://gist.github.com/marenaud/f1c6ed220e63a5939b0d9d32530bebcd):
```
Loading dummy.3ddose
Time to load 3ddose: 440 ms

Loading dummy.bindos
Time to load bindos: 8 ms

Loading dummy.randydos
Time to load randydos: 8 ms
```

randlet commented:
Thanks for running all the numbers...that'll be a nice addition to EGSnrc :)

marenaud commented:
I've updated the pull request and added some documentation / GUI input. I have absolutely no idea what I'm doing for the GUI, but I just copied whatever was there for ihowfarless and it seems to work nicely.

Some things left to discuss would probably be:
- Choice of 0-indexed indices over 1-indexed. 
- I kept num_nonzero in the format even though it's redundant as it can be calculated from the block sizes. I liked having it as a sanity check.

rtownson commented on HEN_HOUSE/omega/progs/gui/dosxyznrc/set_main_inputs_egsnrc.tcl:
```diff
@@ -256,7 +256,7 @@ proc edit_parameters {} {
     frame $w1
     frame $w2
-    foreach j { 5 6 7 8 9 10 12 ihowfarless } {
+    foreach j { 5 6 7 8 9 10 12 ihowfarless ibindos } {
```
Need to define help_text(ibindos) in xyznrc_parameters.tcl

rtownson commented on HEN_HOUSE/doc/src/pirs794-dosxyznrc/pirs794-dosxyznrc.tex:
```diff
@@ -4419,6 +4427,35 @@ \subsection{A Sample {\tt .3ddose} File}
+\indexm{.bindos}
+\indexm{files!.bindos}
+A ``{\tt .bindos}'' file is output instead of a ``{\tt .3ddose}'' file if the
+flag ``{\tt i_bindos}'' is set to 1. The same dose content is contained in each file,
```
Change `{\tt i_bindos}` to `{\tt i\_bindos}`

rtownson commented on HEN_HOUSE/doc/src/pirs794-dosxyznrc/pirs794-dosxyznrc.tex:
```diff
@@ -4419,6 +4427,35 @@ \subsection{A Sample {\tt .3ddose} File}
+\item Voxel boundaries (cm) in z direction($ n_z$ +1 floats)
+\item Number of non-zero dose voxels ($n_{nz}$, 1 int)
+\item Number of voxel index pairs ($n_{b}$, 1 int)
+\item Linearized voxel index pairs indicating the \left[start, end\right) of contiguous non-zero
```
Change `\left[start, end\right)` to `[start, end)`

ftessier commented:
Deferring this to Release 2022. This ought to be merged in develop as soon as possible after Release 2021. @marenaud, Are there any updates at your end: have you or others been using this without issue since implementation?

marenaud commented:
> Deferring this to Release 2022. This ought to be merged in develop as soon as possible after Release 2021. @marenaud, Are there any updates at your end: have you or others been using this without issue since implementation?

Sorry for the late reply. I've used it since implementation with no issues, but it's been a good amount of time since I've generated my last bindos file :(

ftessier commented:
Ok thanks for the update. We will go ahead and merge this in develop after the 2021 release tomorrow!

ftessier commented:
Polished up the commits, empty `git diff` with original branch.

ftessier commented:
Rebased on `develop`.

ftessier commented:
Rebased on develop for merging in release 2022.

ftessier commented:
@mainegra please review when you have a chance.

ftessier commented:
Added author Marc-Andre Renaud in files that were modified to implement the `.bindos` format. Rebased on develop.

ftessier commented:
Add the contributor name in the same commit where the file is mofidied, and squashed the small commit to tweak the dosxyznrc gui menu item strings.
